pyccel/codegen/printing/cucode.py

```diff
@@ -61,14 +61,45 @@ def _print_Module(self, expr):
             else:
                 local_imports += self._print(imp)
 
-        imports = f'{c_headers_imports}\
-                    extern "C"{{\n\
-                    {local_imports}\
-                    }}'
+        imports = f'{c_headers_imports}\n\
+                    {local_imports}\n'
 
         code = f'{imports}\n\
                  {global_variables}\n\
                  {body}\n'
 
         self.exit_scope()
         return code
+
+    def _print_ModuleHeader(self, expr):
+        self.set_scope(expr.module.scope)
+        self._in_header = True
+        name = expr.module.name
+        if isinstance(name, AsName):
+            name = name.name
+        # TODO: Add interfaces
+
+        funcs = ""
+        cuda_headers = ''
+        for f in expr.module.funcs:
+            if not f.is_inline:
+                if 'kernel' in f.decorators:  # Checking for 'kernel' decorator
+                    cuda_headers += self.function_signature(f) + ';\n'
+                else:
+                    funcs += self.function_signature(f) + ';\n'
+        global_variables = ''.join('extern '+self._print(d) for d in expr.module.declarations if not d.variable.is_private)
+        # Print imports last to be sure that all additional_imports have been collected
+        imports = [*expr.module.imports, *self._additional_imports.values()]
+        imports = ''.join(self._print(i) for i in imports)
+
+        self._in_header = False
+        self.exit_scope()
+        imports = f'{cuda_headers}\n\
+                    extern "C"{{\n\
+                    {funcs}\n\
+                    }}'
```

Comment:
Why is a function declaration considered to be an import?

Reply:
I take the old code as an example because I think it's part of Pyccel's philosophy. But for me, it doesn't make any sense. I think next time I should ask you first

Reply:
fixed 9b8627a

Reply:
I agree with this change: https://github.com/pyccel/pyccel-cuda/commit/9b8627a4de86d0336a687678d7058fbb96ebbaa5
But not this one: https://github.com/pyccel/pyccel-cuda/commit/4d85fa542a8e4d10a19672eac7f5c9f0ac9c2c3b

Maybe you meant `function_declarations`?

Reply:
> I take the old code as an example because I think it's part of Pyccel's philosophy.

The old code was:
```python
        imports = f'{c_headers_imports}\
                    extern "C"{{\n\
                    {local_imports}\
                    }}'
```
In this case both `c_headers_imports` and `local_imports` contained prints of `Import`s. The functions are printed at this line:
https://github.com/pyccel/pyccel-cuda/blob/3b0cc72f02c0ceb278bc471fd3878d2937e3ada4/pyccel/codegen/printing/cucode.py#L49

Using this property of  `Module`:
https://github.com/pyccel/pyccel-cuda/blob/3b0cc72f02c0ceb278bc471fd3878d2937e3ada4/pyccel/ast/core.py#L1175-L1180

Your code for `ModuleHeader` doesn't do the same thing. It `_print`s `FunctionDef` headers

Reply:
> But for me, it doesn't make any sense. I think next time I should ask you first

Yes, if you don't understand why something is the way it is it is better to ask rather than copy blindly. As you may have misunderstood and be copying something to somewhere that is not equivalent (as you did here).